### PR TITLE
Issue 365

### PR DESCRIPTION
**Pull request by rpiazza**
#365 Type and property names must begin with a letter 
#368 Custom property names must match [a-z0-9_]

### CODE REVIEW

File: stix2/base.py [hunk 1]
```diff
@@ -76,6 +79,11 @@ def get_required_properties(properties):
 class _STIXBase(Mapping):
     """Base class for STIX object types"""
 
+    def get_class_version(self):
+        module_name = self.__class__.__module__
+        module_parts = module_name.split(".")
+        return module_parts[1]
+
```
Comment (chisholm) [inline stix2/base.py]: Yuuuuck.  An ugly dependency between STIX version and what module a class is in.  I had a similar need in `new_version`, where I needed to know which STIX version the object was so I could use the correct versioning precision rules.  But in that case, the object was fully formed and I could use the `spec_version` property.  Here, it seems you're not so lucky.

Could we just make all 2.1 objects extend a special 2.1 class and use isinstance()?  And the same for 2.0 objects?

File: stix2/base.py [hunk 2]
```diff
@@ -163,6 +171,13 @@ def __init__(self, allow_custom=False, **kwargs):
                 raise ExtraPropertiesError(cls, extra_kwargs)
         if custom_props:
             self._allow_custom = True
+            if self.get_class_version() == "v21":
+                for prop_name, prop_value in custom_props.items():
+                    if not re.match(PREFIX_21_REGEX, prop_name):
+                        raise InvalidValueError(
+                            self.__class__, prop_name,
+                            reason="Property name '%s' must begin with an alpha character." % prop_name,
+                        )
```
Comment (chisholm) [inline stix2/base.py]: I think this only checks custom properties when the object was created with the `custom_properties` parameter.  You can also give the property directly and include `allow_custom=True`, and then the property name wouldn't be checked.  So I think this check needs to be done in a different way.  For example:

```python
f = stix2.v21.File(name="foo.exe", _a_b_=11, allow_custom=True)
```

produces no error.  But:

```python
f = stix2.v21.File(name="foo.exe", custom_properties={"_a_b_": 11})
```

produces an error.

File: stix2/core.py [hunk 3]
```diff
@@ -286,11 +319,12 @@ def _register_observable_extension(
                 ext_type,
             )
     else:  # 2.1+
-        if not re.match(SCO21_EXT_REGEX, ext_type):
+        if not re.match(TYPE_21_REGEX, ext_type):
```
Comment (chisholm) [inline stix2/core.py]: This is for custom SCO extensions: shouldn't the extension name be matched to `EXT_21_REGEX`?

File: stix2/utils.py [hunk 4]
```diff
@@ -25,8 +26,10 @@
 # STIX object properties that cannot be modified
 STIX_UNMOD_PROPERTIES = ['created', 'created_by_ref', 'id', 'type']
 
-TYPE_REGEX = r'^\-?[a-z0-9]+(-[a-z0-9]+)*\-?$'
-SCO21_EXT_REGEX = r'^\-?[a-z0-9]+(-[a-z0-9]+)*\-ext$'
+TYPE_REGEX = re.compile(r'^\-?[a-z0-9]+(-[a-z0-9]+)*\-?$')
+TYPE_21_REGEX = re.compile(r'^\-?([a-z][a-z0-9]*)+(-[a-z0-9]+)*\-?$')
+EXT_21_REGEX = re.compile(r'^\-?([a-z][a-z0-9]*)+(-[a-z0-9]+)*\-ext$')
+PREFIX_21_REGEX = re.compile(r'^[a-z].*')
```
Comment (chisholm) [inline stix2/utils.py]: If the 2.1 rule is that all type names must start with a letter, then `TYPE_21_REGEX` shouldn't start with `^\-?`.  That allows matching strings to start with a hyphen.  Maybe `EXT_21_REGEX` too, if extension names count as "types".

File: stix2/base.py (hunk 1 above)
Comment (rpiazza) [reply stix2/base.py]: @clenk , @chisholm - I like this idea.  What I did was the best I could come up with after discussing it with Chris.  Let's make it independent of this PR, and I can use it once it is implemented and tested

File: stix2/base.py (hunk 2 above)
Comment (rpiazza) [reply stix2/base.py]: I was more concerned about testing with a leading numeric character - which would be impossible to do using the first call above.  I wonder if this should be an illegal property name anyway.

File: stix2/utils.py (hunk 4 above)
Comment (rpiazza) [reply stix2/utils.py]: I didn't add the `\-?`.  It was part of the re before I made any changes.  I guess we should determine if it should be there.  This is related to a comment above...

File: stix2/core.py (hunk 3 above)
Comment (rpiazza) [reply stix2/core.py]: Good catch - fixed

**Comment by rpiazza** [comment]
Doesn't section 3.1 supersede 11.1?

**Comment by emmanvg** [comment]
Actually 11.2 is the section that supersedes it (my bad). I think we can drop my second recommendation for Custom Objects type names. Found the requirement:

- The type property in a Custom Object **MUST** be in ASCII and **MUST** only contain the characters a–z (lowercase ASCII), 0–9, and hyphen (-).
- The type property **MUST NOT** contain a hyphen (-) character immediately following another hyphen (-) character.

File: stix2/base.py (hunk 1 above)
Comment (clenk) [reply stix2/base.py]: Agreed; we should use a 2.1 class with `isinstance()`, and fix it in a subsequent issue/PR.

File: stix2/base.py (hunk 2 above)
Comment (clenk) [reply stix2/base.py]: Yes, we need to enforce this with `allow_custom=True` too.

File: stix2/utils.py (hunk 4 above)
Comment (clenk) [reply stix2/utils.py]: `\-?` shoudn't be there for 2.1.

**Comment by clenk** [comment]
> * I think the type value check can be performed via the `properties.py` module and TypeProperty `class`. This was it is consistent with all other properties allowing: first a spec_version= optional parameter, and a check inside only if the requirement needs to be satisfied for a particular spec_version. This would require all `v21` objects to add the new argument. This helps to remove checking in the module name to determine its version.

I think @chisholm's solution is simpler because we won't have to add a new parameter all over the place. Unless there's a hidden gotcha I'm not thinking of right now.

File: stix2/utils.py [hunk 5]
```diff
@@ -25,8 +26,10 @@
 # STIX object properties that cannot be modified
 STIX_UNMOD_PROPERTIES = ['created', 'created_by_ref', 'id', 'type']
 
-TYPE_REGEX = r'^\-?[a-z0-9]+(-[a-z0-9]+)*\-?$'
-SCO21_EXT_REGEX = r'^\-?[a-z0-9]+(-[a-z0-9]+)*\-ext$'
+TYPE_REGEX = re.compile(r'^\-?[a-z0-9]+(-[a-z0-9]+)*\-?$')
+TYPE_21_REGEX = re.compile(r'([a-z][a-z0-9]*)+(-[a-z0-9]+)*\-?$')
+EXT_21_REGEX = re.compile(r'([a-z][a-z0-9]*)+(-[a-z0-9]+)*\-ext$')
```
Comment (chisholm) [inline stix2/utils.py]: We still need the leading `^` on the regexes to ensure they match the whole string.

**Pull request merged**